src/compute/src/extensions/reduce.rs

```diff
@@ -75,9 +79,50 @@ where
         T2::R: Abelian,
         T2::Batch: Batch,
         L2: FnMut(&K, &[(&V, R)], &mut Vec<(T2::Val, T2::R)>) + 'static,
+        Arranged<G, TraceAgent<T1>>: ArrangementSize,
+        Arranged<G, TraceAgent<T2>>: ArrangementSize,
     {
-        let arranged1 = self.reduce_abelian::<L1, T1>(name1, logic1);
-        let arranged2 = self.reduce_abelian::<L2, T2>(name2, logic2);
+        let arranged1 = self.mz_reduce_abelian::<L1, T1>(name1, logic1);
+        let arranged2 = self.mz_reduce_abelian::<L2, T2>(name2, logic2);
         (arranged1, arranged2)
     }
 }
+
+/// Extension trait for the `reduce_core` differential dataflow method.
+pub(crate) trait MzReduce<G: Scope, K: Data, V: Data, R: Semigroup>:
+    ReduceCore<G, K, V, R>
+where
+    G::Timestamp: Lattice + Ord,
+{
+    /// Applies `reduce` to arranged data, and returns an arrangement of output data.
+    fn mz_reduce_abelian<L, T2>(&self, name: &str, mut logic: L) -> Arranged<G, TraceAgent<T2>>
+    where
+        T2: Trace + TraceReader<Key = K, Time = G::Timestamp> + 'static,
+        T2::Val: Data,
+        T2::R: Abelian,
+        T2::Batch: Batch,
+        L: FnMut(&K, &[(&V, R)], &mut Vec<(T2::Val, T2::R)>) + 'static,
+        Arranged<G, TraceAgent<T2>>: ArrangementSize,
+    {
+        // Allow access to `reduce_core` since we're within Mz's wrapper.
+        #[allow(clippy::disallowed_methods)]
+        self.reduce_core::<_, T2>(name, move |key, input, output, change| {
+            if !input.is_empty() {
+                logic(key, input, change);
+            }
+            change.extend(output.drain(..).map(|(x, d)| (x, d.negate())));
+        })
```

Comment:
Good question! It's probably better to call `reduce_abelian`. Removing the `consolidate` call eliminates redundant work, because the reduction thinker immediately consolidates the changes after the closure returns. We should fix this in Differential!

Reply:
Ah, that makes sense. Agreed that it's better to call `reduce_abelian` here and open a PR for DD!